### PR TITLE
Implementation: Support for Bearer token auth between Service Catalog and brokers

### PR DESCRIPTION
**Pull request by nilebox**
Bearer token auth support **code** implementation based on the ideas from two proposals:
- Proposal for opaque bearer token auth https://github.com/kubernetes-incubator/service-catalog/pull/1044
- Proposal for additional auth support https://github.com/kubernetes-incubator/service-catalog/pull/1037

Note that there are some differences in the implementation compared to the proposals mentioned above.

I went without extra abstraction of `AuthInterface` for now, just extended existing `AuthConfig` type. While it's not as flexible as `AuthInterface` approach, it is much easier to reason about, and does not have to provide an abstraction for other auth techniques, such as x509 mentioned in https://github.com/kubernetes-incubator/service-catalog/pull/1037#discussion_r127596077.

### CODE REVIEW

File: vendor/github.com/pmorie/go-open-service-broker-client/v2/client.go [hunk 1]
```diff
@@ -1,3 +1,4 @@
+// TODO (nilebox): Revert and update the dependency after PR merged: https://github.com/pmorie/go-open-service-broker-client/pull/50
```
Comment (nilebox) [inline vendor/github.com/pmorie/go-open-service-broker-client/v2/client.go]: This is a temporary copy-pasta until https://github.com/pmorie/go-open-service-broker-client/pull/50 gets merged.

File: vendor/github.com/pmorie/go-open-service-broker-client/v2/interface.go [hunk 2]
```diff
@@ -1,3 +1,4 @@
+// TODO (nilebox): Revert and update the dependency after PR merged: https://github.com/pmorie/go-open-service-broker-client/pull/50
```
Comment (nilebox) [inline vendor/github.com/pmorie/go-open-service-broker-client/v2/interface.go]: This is a temporary copy-pasta until https://github.com/pmorie/go-open-service-broker-client/pull/50 gets merged.

File: cmd/apiserver/app/server/run_server.go [hunk 3]
```diff
@@ -183,7 +183,7 @@ func (c checkEtcdConnectable) Check(_ *http.Request) error {
 	serverReachable, err := preflight.EtcdConnection{ServerList: c.ServerList}.CheckEtcdServers()
 
 	if err != nil {
-		glog.Error("etcd checker failed with err: %v", err)
+		glog.Errorf("etcd checker failed with err: %v", err)
```
Comment (nilebox) [inline cmd/apiserver/app/server/run_server.go]: @MHBauer fix for your change in https://github.com/kubernetes-incubator/service-catalog/pull/1039
not sure how did it pass through Travis verification build:
```
Running golint and go vet:
docker run --rm -v /Users/nislamov/go/src/github.com/kubernetes-incubator/service-catalog:/go/src/github.com/kubernetes-incubator/service-catalog -v /Users/nislamov/go/src/github.com/kubernetes-incubator/service-catalog/.pkg:/go/pkg scbuildimage go vet ./cmd/... ./contrib/... ./pkg/... ./plugin/... ./test/...
cmd/apiserver/app/server/run_server.go:186: possible formatting directive in Error call
exit status 1
make: *** [verify] Error 1
```

Comment (ash2k) [inline cmd/apiserver/app/server/run_server.go]: I haven't looked at the details of the call chain, but shouldn't it be either logging OR returning that error? That caller will probably also log this so I'll end up in the logs twice, causing confusion. Why not `return fmt.Errorf(....)`?

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -373,37 +374,64 @@ func (c *controller) getServiceClassPlanAndBrokerForBinding(instance *v1alpha1.I
...
+	case v1alpha1.SecretTypeBearer:
+		bearerConfig, err := getBearerConfig(authSecret)
+		if err != nil {
+			return nil, err
+		}
+		return &osb.AuthConfig{
+			BearerConfig: bearerConfig,
+		}, nil
+	default:
+		return nil, fmt.Errorf("Unsupported secret type: %s", broker.Spec.AuthInfo.SecretType)
```
Comment (ash2k) [inline pkg/controller/controller.go]: nit: the convention is to use lowercase letter i.e. `unsupported ....`.

File: cmd/apiserver/app/server/run_server.go (hunk 3 above)
Comment (nilebox) [inline cmd/apiserver/app/server/run_server.go]: @ash2k better ask @MHBauer in a separate thread, I just fixed the formatting error.

File: contrib/examples/apiserver/broker.yaml [hunk 5]
```diff
@@ -8,6 +8,6 @@ spec:
   # service-catalog will use the contents of the secret. The secret should have "username"
   # and "password" keys
   authInfo:
-    basicAuthSecret:
+    authSecret:
```
Comment (nilebox) [inline contrib/examples/apiserver/broker.yaml]: `authInfo.authSecret` looks redundant.
Shall I rename it to just `secret` or `secretRef`?

Comment (nilebox) [inline contrib/examples/apiserver/broker.yaml]: renamed to `secretRef`

File: cmd/apiserver/app/server/run_server.go (hunk 3 above)
Comment (MHBauer) [inline cmd/apiserver/app/server/run_server.go]: Formatter looks wrong to me too. Agree it should be `Errorf`. Very not sure how it got through the build. I know at some points we have had verification that didn't fail the build.

I don't think the `/healthz` healthchecker endpoint logs anything internally, so I'm raising it to the operator if they ever check the logs that etcd might be down or not connectable.

This code gets called whenever `GET /healthz` happens. Usually this is done during the liveness/readiness probes by k8s infra. So it could be called repeatedly over a long period of time, but as long as it never fails, nothing should be logged. We could maybe change the level of the info at the top to be a lower level.

Comment (nilebox) [inline cmd/apiserver/app/server/run_server.go]: @MHBauer can you raise a separate PR with fix for this? This PR could take long time to be approved and merged.

File: pkg/apis/servicecatalog/testing/fuzzer.go [hunk 6]
```diff
@@ -240,6 +240,16 @@ func FuzzerFor(t *testing.T, version schema.GroupVersion, src rand.Source) *fuzz
 			sp.AlphaInstanceCreateParameterSchema = metadata
 			sp.AlphaInstanceUpdateParameterSchema = metadata
 		},
+		func(bs *servicecatalog.BrokerSpec, c fuzz.Continue) {
+			c.FuzzNoCustom(bs)
+			// Don't allow the SecretName to be an empty string because
```
Comment (bmelville) [inline pkg/apis/servicecatalog/testing/fuzzer.go]: SecretType

File: cmd/apiserver/app/server/run_server.go (hunk 3 above)
Comment (MHBauer) [inline cmd/apiserver/app/server/run_server.go]: #1056

File: pkg/apis/servicecatalog/testing/fuzzer.go (hunk 6 above)
Comment (nilebox) [inline pkg/apis/servicecatalog/testing/fuzzer.go]: Fixed

**Comment by nilebox** [comment]
> I'm wondering, is there no controller_test.go with tests for the bulk of the code you've modified?

@bmelville I'm happy to add one if the code itself looks good.
I don't see a separate test for (existing) basic auth either, though.

File: pkg/controller/controller_broker_test.go [hunk 7]
```diff
@@ -372,33 +373,141 @@ func TestReconcileBrokerZeroServices(t *testing.T) {
...
+			authInfo:      bearerAuthInfo,
+			secret:        nil,
+			shouldSucceed: false,
+		},
+	}
 
-	if err := testController.reconcileBroker(broker); err == nil {
-		t.Fatal("Should have failed to get the auth for the broker.")
+	for _, tc := range cases {
+		testReconcileBrokerWithAuth(t, tc.authInfo, tc.secret, tc.shouldSucceed)
```
Comment (ash2k) [inline pkg/controller/controller_broker_test.go]: Consider using subtests `t.Run(...)`

File: pkg/controller/controller_broker_test.go [hunk 8]
```diff
@@ -372,33 +373,144 @@ func TestReconcileBrokerZeroServices(t *testing.T) {
 	}
 }
 
-func TestReconcileBrokerWithAuthError(t *testing.T) {
-	fakeKubeClient, fakeCatalogClient, fakeBrokerClient, testController, _ := newTestController(t, fakeosb.FakeClientConfiguration{})
+func TestReconcileBrokerWithAuth(t *testing.T) {
```
Comment (nilebox) [inline pkg/controller/controller_broker_test.go]: @bmelville please review the unit test for basic and bearer auth below.

File: contrib/examples/apiserver/broker.yaml [hunk 9]
```diff
@@ -8,6 +8,7 @@ spec:
   # service-catalog will use the contents of the secret. The secret should have "username"
   # and "password" keys
   authInfo:
-    basicAuthSecret:
+    secretRef:
```
Comment (arschles) [inline contrib/examples/apiserver/broker.yaml]: @nilebox would it be possible to eliminate the need to use `secretType`? When I put in the code for `authInfo`, I allowed for other keys besides `basicAuthSecret`. For example:

```yaml
authInfo:
    bearerTokenAuthSecret:
        namespace: some-namespace
        name: secret-name
```

Either way is fine with me, I'm just looking for an explanation for why the change.

File: pkg/apis/servicecatalog/types.go [hunk 10]
```diff
@@ -56,15 +56,40 @@ type BrokerSpec struct {
...
+	// BasicAuthUsernameKey is the key of the username for SecretTypeBasicAuth secrets
+	BasicAuthUsernameKey = "username"
+	// BasicAuthPasswordKey is the key of the password or token for SecretTypeBasicAuth secrets
+	BasicAuthPasswordKey = "password"
+
+	// SecretTypeBearer contains data needed for bearer token authentication.
+	//
+	// Required field:
+	// - Secret.Data["token"] - bearer token for authentication
+	SecretTypeBearer BrokerAuthSecretType = "bearer"
```
Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: Naming suggestion: `SecretTypeBearerTokenAuth`, so that it's more descriptive and is consistent with the `Auth` suffix above. I won't push any further than this on naming; feel free to ignore if you don't like it 😁

File: pkg/apis/servicecatalog/types.go [hunk 11]
```diff
@@ -56,15 +56,40 @@ type BrokerSpec struct {
...
+	SecretTypeBasicAuth BrokerAuthSecretType = "basic"
+	// BasicAuthUsernameKey is the key of the username for SecretTypeBasicAuth secrets
+	BasicAuthUsernameKey = "username"
+	// BasicAuthPasswordKey is the key of the password or token for SecretTypeBasicAuth secrets
+	BasicAuthPasswordKey = "password"
+
+	// SecretTypeBearer contains data needed for bearer token authentication.
+	//
+	// Required field:
+	// - Secret.Data["token"] - bearer token for authentication
```
Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: Can you add this information also to [these comments](https://github.com/kubernetes-incubator/service-catalog/blob/master/contrib/examples/apiserver/broker.yaml#L7)?

**Comment by pmorie** [comment]
I'd like to reserve the chance to add the second LGTM here

File: contrib/examples/apiserver/broker.yaml (hunk 9 above)
Comment (pmorie) [inline contrib/examples/apiserver/broker.yaml]: this is an API break, so we should let people know that this is going to change.

File: pkg/apis/servicecatalog/types.go (hunk 10 above)
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: I also support `SecretTypeBearerTokenAuth`

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 12]
```diff
@@ -57,15 +57,40 @@ type BrokerSpec struct {
...
-// set on it.
 type BrokerAuthInfo struct {
-	// BasicAuthSecret is a reference to a Secret containing auth information the
-	// catalog should use to authenticate to this Broker using basic auth.
-	BasicAuthSecret *v1.ObjectReference `json:"basicAuthSecret,omitempty"`
+	// SecretRef is a reference to a Secret containing auth information the
+	// catalog should use to authenticate to this Broker using corresponding auth.
+	SecretRef *v1.ObjectReference `json:"secretRef,omitempty"`
+	// Used to facilitate programmatic handling of auth secret data.
+	// Defaults to "basic"
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Nit: this should end with a period

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 13]
```diff
@@ -57,15 +57,40 @@ type BrokerSpec struct {
...
+	// +optional
+	SecretType BrokerAuthSecretType `json:"secretType,omitempty"`
 }
 
+// BrokerAuthSecretType is an identifier of Secret type to be used for authentication
```
Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Nit: period

File: pkg/controller/controller_actions_test.go [hunk 14]
```diff
@@ -167,3 +168,23 @@ func checkCatalogClientActions(actual []testing.Action, expected []catalogClient
...
 	return nil
 }
+
+type brokerClientAction struct {
+	actionType fakeosb.ActionType
+}
+
+// checkBrokerClientActions is the utility function for checking actions returned by
+// the broker client
+func checkBrokerClientActions(actual []fakeosb.Action, expected []brokerClientAction) error {
```
Comment (pmorie) [inline pkg/controller/controller_actions_test.go]: I'm a little worried that we have a few different ways of doing this type of check proliferating - did you see `assertGetCatalog` in `controller_test.go`?

File: test/integration/clientset_test.go [hunk 15]
```diff
@@ -287,12 +287,12 @@ func testBrokerClient(sType server.StorageType, client servicecatalogclient.Inte
...
 	}
 
-	brokerServer.Spec.AuthInfo = &v1alpha1.BrokerAuthInfo{BasicAuthSecret: authSecret}
+	brokerServer.Spec.AuthInfo = &v1alpha1.BrokerAuthInfo{SecretRef: authSecret}
 
 	brokerUpdated, err := brokerClient.Update(brokerServer)
 	if nil != err ||
-		"test-namespace" != brokerUpdated.Spec.AuthInfo.BasicAuthSecret.Namespace ||
-		"test-name" != brokerUpdated.Spec.AuthInfo.BasicAuthSecret.Name {
+		"test-namespace" != brokerUpdated.Spec.AuthInfo.SecretRef.Namespace ||
```
Comment (pmorie) [inline test/integration/clientset_test.go]: How do you feel about introducing some constants here?

File: contrib/examples/apiserver/broker.yaml (hunk 9 above)
Comment (nilebox) [inline contrib/examples/apiserver/broker.yaml]: @arschles I prefer to keep `secretType` instead of duplicating the API and code, but it's up to the discussion of course.
Even if we were to eliminate the `secretType` property, I would still make a breaking change:
```yaml
authInfo:
    basicAuth:
        secret:
          namespace: some-namespace
          name: secret-name
```
and
```yaml
authInfo:
    bearerTokenAuth:
        secret:
          namespace: some-namespace
          name: secret-name
```
to decouple auth mode from secret.

File: pkg/apis/servicecatalog/types.go (hunk 11 above)
Comment (nilebox) [inline pkg/apis/servicecatalog/types.go]: @arschles the example you referenced uses Basic auth, I can't add information about Bearer token auth there.

File: contrib/examples/apiserver/broker.yaml (hunk 9 above)
Comment (nilebox) [inline contrib/examples/apiserver/broker.yaml]: Current approach also replicates the way `Secret.Type` is implemented: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/types.go#L3841

File: pkg/controller/controller_actions_test.go (hunk 14 above)
Comment (nilebox) [inline pkg/controller/controller_actions_test.go]: @pmorie I didn't, thanks for the suggestion

File: test/integration/clientset_test.go (hunk 15 above)
Comment (nilebox) [inline test/integration/clientset_test.go]: Well, those inline strings were all around the place before I added my test, just trying to be consistent.
If you insist on moving all of them into constants, I will :)

File: pkg/apis/servicecatalog/types.go (hunk 11 above)
Comment (nilebox) [inline pkg/apis/servicecatalog/types.go]: @arschles added comment to `broker.yaml` about bearer token support as an alternative auth.

**Comment by nilebox** [comment]
@pmorie finished making changes to switch to the union type, please review.

File: pkg/apis/servicecatalog/v1alpha1/types.go [hunk 16]
```diff
@@ -57,15 +57,44 @@ type BrokerSpec struct {
...
-// Note that we currently restrict a single broker to have only one of these fields
-// set on it.
 type BrokerAuthInfo struct {
-	// BasicAuthSecret is a reference to a Secret containing auth information the
-	// catalog should use to authenticate to this Broker using basic auth.
-	BasicAuthSecret *v1.ObjectReference `json:"basicAuthSecret,omitempty"`
+	// Basic provides configuration for basic authentication
+	Basic *BasicAuthConfig `json:"basic,omitempty"`
+	// Bearer provides configuration for bearer token authentication
+	Bearer *BearerTokenAuthConfig `json:"bearer,omitempty"`
```
Comment (vaikas-google) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Could we not call this BearerTokenAuthConfig and call it something like:
AuthenticationHeader

That way it could be used for something else in the future. If I understand this PR correctly, refresh -> bearer token happens offline somewhere else? So, would be good to at least walk through and see what that process looks like. Is it assumed that offline process handles the conversion [aka, possibly continously refreshing [unnecessarily if no calls to broker are made]]. Is there no plan to store the refresh token as part of the Service Catalog, and having a way to convert that into a bearer token?

Comment (liggitt) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: bearer token is the correct name, and refers to the way the token is *used* (sent in an `Authorization: Bearer <token>` header), not obtained

Comment (vaikas-google) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Yeah, I understand that, but since it's opaque, why is Bearer special? If we just make the entire header opaque, you could support things like JWT without making changes to the api.

Comment (vaikas-google) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Obviously, the header wouldn't then tack on the Bearer there, what I'm saying is that if we specify opaque things, why slap the Bearer there, make the entire Authorization header opaque. I don't see any value in tacking on the the Bearer part.

Comment (liggitt) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: making the thing obtaining this token form a secret containing the value `Bearer <token>` is odd.

Comment (pmorie) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Seems like entirely opaque auth header is not usable with k8s service account tokens.  I'm not against fully opaque controller over the auth header value, but i think we should also have something that works correctly with k8s service account tokens.

**Comment by vaikas-google** [comment]
Then how about a type field? It just feels clunky to bake the type into the
name since for any new method we have to add effectively code to the
controller to simply add a strong that encodes the type of token.

**Comment by liggitt** [comment]
> Then how about a type field? It just feels clunky to bake the type into the name since for any new method we have to add effectively code to the controller to simply add a strong that encodes the type of token.

basic and bearer auth schemes account for the majority of the static credentials you are going to encounter. I think it is worth integration with those in a first-class way. If you run into other use cases, a fully generic Authorization header option could be added later, but I'm skeptical that will be necessary (most others I'm aware of have a dynamic component, like negotiate, etc)

**Comment by vaikas-google** [comment]
Bearer is not static though? It's only static because something else
exchanges the refresh token for a bearer token?

**Comment by liggitt** [comment]
> Bearer is not static though? It's only static because something else exchanges the refresh token for a bearer token?

By static, I mean not tied to a particular request, as opposed to a `Authorization: Negotiate` authorization that requires sending the same request multiple times, changing what is sent in response to a `WWW-Authenticate` challenge from the server

File: pkg/apis/servicecatalog/v1alpha1/types.go (hunk 16 above)
Comment (nilebox) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: @vaikas-google FYI we have discussed this question at the today's call, and the decision was to stick with the current proposal (explicitly support `bearer` instead of adding support for arbitrary `Authorization:` header with any prefix).

Comment (vaikas-google) [inline pkg/apis/servicecatalog/v1alpha1/types.go]: Ok, well then :)

File: pkg/apis/servicecatalog/types.go [hunk 17]
```diff
@@ -56,15 +56,44 @@ type BrokerSpec struct {
...
 // the the service catalog and brokers may support, according to the OpenServiceBroker API
 // specification (https://github.com/openservicebrokerapi/servicebroker/blob/master/spec.md).
-//
-// Note that we currently restrict a single broker to have only one of these fields
-// set on it.
 type BrokerAuthInfo struct {
-	// BasicAuthSecret is a reference to a Secret containing auth information the
-	// catalog should use to authenticate to this Broker using basic auth.
-	BasicAuthSecret *v1.ObjectReference
+	// Basic provides configuration for basic authentication
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: Use periods in godoc comments like this.

File: pkg/apis/servicecatalog/types.go [hunk 18]
```diff
@@ -56,15 +56,44 @@ type BrokerSpec struct {
...
+	Basic *BasicAuthConfig
+	// Bearer provides configuration for bearer token authentication
+	Bearer *BearerTokenAuthConfig
 }
 
+// BasicAuthConfig provides config for the basic authentication.
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: s/the//

File: pkg/apis/servicecatalog/types.go [hunk 19]
```diff
@@ -56,15 +56,44 @@ type BrokerSpec struct {
...
+}
+
+// BearerTokenAuthConfig provides config for the bearer token authentication.
+type BearerTokenAuthConfig struct {
+	// SecretRef is a reference to a Secret containing information the
+	// catalog should use to authenticate to this Broker.
+	//
+	// Required field:
+	// - Secret.Data["token"] - bearer token for authentication
+	SecretRef *v1.ObjectReference
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: This should be a `LocalObjectReference`

File: pkg/apis/servicecatalog/types.go [hunk 20]
```diff
@@ -56,15 +56,44 @@ type BrokerSpec struct {
...
+	// SecretRef is a reference to a Secret containing information the
+	// catalog should use to authenticate to this Broker.
+	//
+	// Required at least one of the fields:
+	// - Secret.Data["username"] - username used for authentication
+	// - Secret.Data["password"] - password or token needed for authentication
+	SecretRef *v1.ObjectReference
+}
+
+// BearerTokenAuthConfig provides config for the bearer token authentication.
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: I think we should be clear about exactly what the behavior this has.  I'd say the following:

```go
// BearerTokenAuthConfig provides configuration to send an opaque value as a bearer token.
// The value is referenced from the 'token' field of the given secret.  This value should only
// contain the token value and not the `Bearer` scheme.
```

File: pkg/controller/controller.go [hunk 21]
```diff
@@ -369,41 +370,71 @@ func (c *controller) getServiceClassPlanAndBrokerForBinding(instance *v1alpha1.I
 // Broker utility methods - move?
 
 // getAuthCredentialsFromBroker returns the auth credentials, if any,
-// contained in the secret referenced in the Broker's AuthSecret field, or
-// returns an error. If the AuthSecret field is nil, empty values are
+// contained in the secret referenced in the Broker's SecretRef field, or
```
Comment (pmorie) [inline pkg/controller/controller.go]: I don't think this comment is accurate anymore.

**Comment by pmorie** [comment]
@nilebox you can eliminate the OSB bump in this PR now.

**Comment by nilebox** [comment]
@pmorie I think this PR is ready to be merged now.

File: pkg/apis/servicecatalog/types.go [hunk 22]
```diff
@@ -56,15 +56,51 @@ type BrokerSpec struct {
...
 
+// BasicAuthConfig provides config for the basic authentication.
+type BasicAuthConfig struct {
+	// SecretRef is a reference to a Secret containing information the
+	// catalog should use to authenticate to this Broker.
+	//
+	// Required at least one of the fields:
+	// - Secret.Data["username"] - username used for authentication
+	// - Secret.Data["password"] - password or token needed for authentication
+	SecretRef *v1.LocalObjectReference
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: I am sorry, I think I gave you wrong advice.  This needs to be an ObjectReference because it is from a global to a namespaced context.

File: pkg/apis/servicecatalog/types.go [hunk 23]
```diff
@@ -56,15 +56,51 @@ type BrokerSpec struct {
...
+}
+
+// BearerTokenAuthConfig provides config for the bearer token authentication.
+type BearerTokenAuthConfig struct {
+	// SecretRef is a reference to a Secret containing information the
+	// catalog should use to authenticate to this Broker.
+	//
+	// Required field:
+	// - Secret.Data["token"] - bearer token for authentication
+	SecretRef *v1.LocalObjectReference
```
Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: Same thing here - needs to be an ObjectReference since it's a global -> namespaced reference.  My bad!

File: pkg/apis/servicecatalog/types.go (hunk 22 above)
Comment (nilebox) [inline pkg/apis/servicecatalog/types.go]: > it is from a global to a namespaced context

what do you mean global? Broker resource is namespaced, right (just like any other Kubernetes resource)? Do we have a use case for referencing a secret in a different namespace than the Broker resource?
